core/trino-main/src/main/java/io/trino/execution/scheduler/EventDrivenFaultTolerantQueryScheduler.java

```diff
@@ -0,0 +1,2103 @@
+            this.plan = requireNonNull(plan, "plan is null");
+
+            planInTopologicalOrder = sortPlanInTopologicalOrder(plan);
+        }
+
+        public void run()
+        {
+            checkState(!started, "already started");
+            started = true;
+
+            queryStateMachine.addStateChangeListener(state -> {
+                if (state.isDone()) {
+                    eventQueue.add(Event.WAKE_UP);
+                }
+            });
+
+            Optional<Throwable> failure = Optional.empty();
+            try {
+                if (schedule()) {
+                    while (processEvents()) {
+                        if (delayStopwatch.isRunning()) {
+                            verify(delaySchedulingDurationInMillis.isPresent(), "delaySchedulingDurationInMillis must be present");
+                            long remainingDelay = delaySchedulingDurationInMillis.getAsLong() - delayStopwatch.elapsed(MILLISECONDS);
+                            if (remainingDelay > 0) {
+                                continue;
+                            }
+                        }
+                        if (!schedule()) {
+                            break;
+                        }
+                    }
+                }
+            }
+            catch (Throwable t) {
+                failure = Optional.of(t);
+            }
+
+            for (StageExecution execution : stageExecutions.values()) {
+                failure = closeAndAddSuppressed(failure, execution::abort);
+            }
+            for (NodeLease nodeLease : nodeAcquisitions.values()) {
+                failure = closeAndAddSuppressed(failure, nodeLease::release);
+            }
+            nodeAcquisitions.clear();
+            failure = closeAndAddSuppressed(failure, nodeAllocator);
+
+            failure.ifPresent(queryStateMachine::transitionToFailed);
+        }
+
+        private Optional<Throwable> closeAndAddSuppressed(Optional<Throwable> existingFailure, Closeable closeable)
+        {
+            try {
+                closeable.close();
+            }
+            catch (Throwable t) {
+                if (existingFailure.isEmpty()) {
+                    return Optional.of(t);
+                }
+                if (existingFailure.get() != t) {
+                    existingFailure.get().addSuppressed(t);
+                }
+            }
+            return existingFailure;
+        }
+
+        private boolean processEvents()
+        {
+            try {
+                Event event = eventQueue.poll(1, MINUTES);
+                if (event == null) {
+                    return true;
+                }
+                eventBuffer.add(event);
+            }
+            catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+                throw new RuntimeException(e);
+            }
+
+            while (true) {
+                eventQueue.drainTo(eventBuffer, EVENT_BUFFER_CAPACITY - eventBuffer.size());
```

Comment:
This is an efficiency optimization to allow pulling multiple events from the queue in one shot. The `eventBuffer` is kept as a filed to avoid allocating it every time.